crawlerr/services/strategies/web_strategy.py

```
@@ -0,0 +1,402 @@
+                and depth < request.max_depth
+                and len(crawled_urls) < request.max_pages
+            ):
+                current_level_urls = urls_to_crawl.copy()
+                urls_to_crawl.clear()
+                next_level_urls = []
+
+                # Process URLs in current level with limited concurrency
+                semaphore = asyncio.Semaphore(min(settings.max_concurrent_crawls, 5))
+
+                async def crawl_single_url(url: str) -> PageContent | Exception:
+                    async with semaphore:
```

Ruff (0.12.2)

Function definition does not bind loop variable `semaphore`

(B023)